File: 1592. Rearrange Spaces Between Words/solution.java
class Solution {
  public String reorderSpaces(String text) {
    String[] words = text.trim().split("\\s+");
    int spaces = 0;
    for (int index = 0; index < text.length(); index++){
      Character letter = text.charAt(index);
      if (letter == ' '){
        spaces += 1;      
      }
    }
    
    int gap = 0;
    if (words.length != 1){
      gap = spaces / (words.length - 1);    
    }
    
    int extraSpaces = spaces - gap * (words.length - 1);
    
    String spacesBetween = "";
    for (int index = 0; index < gap; index++){
      spacesBetween += " ";
    }
    
    String endingSpaces = "";
    for (int index = 0; index < extraSpaces; index++){
      endingSpaces += " ";
    }

    String result = "";
    for (int index = 0; index < words.length; index++){
      String word = words[index];
      result += word;
      if (index != words.length - 1){
        result += spacesBetween;
      }
      else{
        result += endingSpaces;
      }
    }

    return result;
  }
}

File: 1592. Rearrange Spaces Between Words/solution.cpp
class Solution {
public:
  string reorderSpaces(string text) {
    vector<string> words;
    string word = "";
    int spaces = 0;
    for (int index = 0; index < text.size(); index++){
      char letter = text[index];
      if (letter != ' '){
        word += letter;
      }
      else{
        spaces += 1;
      }
      if ((letter == ' ' || index == text.size() - 1) && word.size() > 0){
        words.push_back(word);
        word = "";
      }
    }

    int gap = 0;
    if (words.size() != 1){
      gap = spaces / (words.size() - 1);    
    }
    
    int extraSpaces = spaces - gap * (words.size() - 1);
    
    string spacesBetween = "";
    for (int index = 0; index < gap; index++){
      spacesBetween += " ";
    }
    
    string endingSpaces = "";
    for (int index = 0; index < extraSpaces; index++){
      endingSpaces += " ";
    }

    string result = "";
    for (int index = 0; index < words.size(); index++){
      string word = words[index];
      result += word;
      if (index != words.size() - 1){
        result += spacesBetween;
      }
      else{
        result += endingSpaces;
      }
    }

    return result;
  }
};